runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/GrpcWindmillServer.java

```diff
@@ -1525,8 +1525,8 @@ public boolean hasMoreElements() {
 
                 try {
                   blockedStartMs.set(Instant.now().getMillis());
-                  current = queue.take();
-                  if (current != POISON_PILL) {
+                  current = queue.poll(180, TimeUnit.SECONDS);
+                  if (current != null && current != POISON_PILL) {
```

Comment:
If `current == null` we could fall through to the `Got poison pill but stream is not done.` exception. Shouldn't we introduce a separate case for the timeout?

Reply:
I don't think we need a separate exception - i just updated the error message to indicate either case was possible